lib/puppet/environments.rb

```diff
@@ -476,6 +484,20 @@ def get_conf(name)
       @loader.get_conf(name)
     end
 
+    # Guard an environment so it can't be evicted while it's in use. The method
+    # may be called multiple times, provided it is unguarded the same number of
+    # times. If you call this method, you must call `unguard` in an ensure block.
+    def guard(name)
+      entry = get_entry(name, false)
```

Comment:
Why do we not want to `check_expired` for these? Did that cause a specific bug?

Reply:
It was overkill in the case of `guard`, for example in the old PR we loaded/cleared once in:

https://github.com/puppetlabs/puppet/pull/8673/files#diff-e665c46dc177d4c7e13d6ff5f107c10926a471b9913b1bdd9bdfea63a31e0148R816

And then again when guarding the environment: 

https://github.com/puppetlabs/puppet/pull/8673/files#diff-e665c46dc177d4c7e13d6ff5f107c10926a471b9913b1bdd9bdfea63a31e0148R817

So there are 2 "created" environments, but it's not clear which methods caused the environment to be created twice.

Also there was some complexity/magic with `unguard`. It would try to clear expired environments, but since the environment was still guarded, it didn't clear itself (fortunately). But by decoupling guarding and expiration, the complexity goes away.